src/subscriptions/SubscriptionManager.h

```diff
@@ -0,0 +1,290 @@
+
+class Subscription
+{
+    boost::asio::io_context::strand strand_;
+    std::unordered_set<std::shared_ptr<WsBase>> subscribers_ = {};
+
+    void
+    sendAll(std::string const& pubMsg);
+
+public:
+    static std::unique_ptr<Subscription>
+    make_Subscription(boost::asio::io_context& ioc) {
+        return std::make_unique<Subscription>(ioc);
+    }
+
+    Subscription() = delete;
+    Subscription(Subscription&) = delete;
+    Subscription(Subscription&&) = delete;
+
+    explicit 
+    Subscription(boost::asio::io_context& ioc) : strand_(ioc)
+    {    
+    }
+
+    ~Subscription() = default;
+
+    void
+    subscribe(std::shared_ptr<WsBase> const& session);
+
+    void
+    unsubscribe(std::shared_ptr<WsBase> const& session);
+
+    void
+    publish(std::string const& message);
+};
+
+class AccountSubscription
+{
+    boost::asio::io_context::strand strand_;
+    std::unordered_map<
+        ripple::AccountID,
+        std::unordered_set<std::shared_ptr<WsBase>>> subscribers_ = {};
+
+    void
+    sendAll(std::string const& pubMsg, ripple::AccountID const& account);
+
+public:
+    static std::unique_ptr<AccountSubscription>
+    make_Subscription(boost::asio::io_context& ioc) {
+        return std::make_unique<AccountSubscription>(ioc);
+    }
+
+    AccountSubscription() = delete;
+    AccountSubscription(AccountSubscription&) = delete;
+    AccountSubscription(AccountSubscription&&) = delete;
+
+    explicit 
+    AccountSubscription(boost::asio::io_context& ioc) : strand_(ioc) 
+    {
+    }
+
+    ~AccountSubscription() = default;
+
+    void
+    subscribe(
+        std::shared_ptr<WsBase> const& session,
+        ripple::AccountID const& account);
+
+    void
+    unsubscribe(
+        std::shared_ptr<WsBase> const& session,
+        ripple::AccountID const& account);
+
+    void
+    publish(
+        std::string const& message,
+        ripple::AccountID const& account);
+};
+
+class BookSubscription
+{
+
+    boost::asio::io_context::strand strand_;
+    std::unordered_map<
+        ripple::Book,
+        std::unordered_set<std::shared_ptr<WsBase>>> subscribers_ = {};
+
+    void
+    sendAll(std::string const& pubMsg, ripple::Book const& book);
+
+public:
+    static std::unique_ptr<BookSubscription>
+    make_Subscription(boost::asio::io_context& ioc) {
+        return std::make_unique<BookSubscription>(ioc);
+    }
+
+    BookSubscription() = delete;
+    BookSubscription(BookSubscription&) = delete;
+    BookSubscription(BookSubscription&&) = delete;
+
+    explicit 
+    BookSubscription(boost::asio::io_context& ioc) : strand_(ioc) 
+    {
+    }
+
+    ~BookSubscription() = default;
+
+    void
+    subscribe(
+        std::shared_ptr<WsBase> const& session,
+        ripple::Book const& book);
+
+    void
+    unsubscribe(
+        std::shared_ptr<WsBase> const& session,
+        ripple::Book const& book);
+
+    void
+    publish(
+        std::string const& message,
+        ripple::Book const& book);
+};
```

Comment:
Let's make these a template class. `BookSubscription` and `AccountSubscription` are easily templatized, where the map key is the template parameter. They can call template helper functions, which `Subscription` would also call. Consider something like this:
```
template <class T>
void
sendAll(std::string const& pubMsg, T& subscribers)
{
    for (auto it = subscribers.begin(); it != subscribers.end();)
    {
        auto& session = *it;
        if (session->dead())
        {
            it = subscribers.erase(it);
        }
        else
        {
            session->send(pubMsg);
            ++it;
        }
    }
}

template <class T>
void
subscribe(std::shared_ptr<WsBase> const& session, T& subscribers, boost::asio::io_context::strand& strand)
{
    boost::asio::post(strand, [subscribers, session](){
        subscribers.emplace(session);
    });
}

template <class T>
void
unsubscribe(std::shared_ptr<WsBase> const& session, T& subscribers, boost::asio::io_context::strand& strand)
{
    boost::asio::post(strand, [subscribers, session](){
        subscribers.erase(session);
    });
}

```

Does this make sense?

Reply:
Yup, implemented something similar and testing it out.